codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java

```diff
@@ -0,0 +1,347 @@
+            promise.setFailure(
+                    connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+        }
+        return promise;
+    }
+
+    @Override
+    public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data,
+                                   int padding,
+                                   boolean endOfStream, ChannelPromise promise) {
+        if (existingStream(streamId)) {
+            return super.writeData(ctx, streamId, data, padding, endOfStream, promise);
+        }
+        PendingStream pendingStream = pendingStreams.get(streamId);
+        if (pendingStream != null) {
+            pendingStream.frames.add(new DataFrame(data, padding, endOfStream, promise));
+        } else {
+            ReferenceCountUtil.safeRelease(data);
+            promise.setFailure(
+                    connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+        }
+        return promise;
+    }
+
+    @Override
+    public ChannelFuture writeSettingsAck(ChannelHandlerContext ctx, ChannelPromise promise) {
+        receivedSettings = true;
+        ChannelFuture future = super.writeSettingsAck(ctx, promise);
+        // After having received a SETTINGS frame, the maximum number of concurrent streams
+        // might have changed. So try to create some buffered streams.
+        tryCreatePendingStreams();
+        return future;
+    }
+
+    @Override
+    public void close() {
+        super.close();
+        cancelPendingStreams();
+    }
+
+    private void tryCreatePendingStreams() {
+        while (!pendingStreams.isEmpty() && canCreateStream()) {
+            Map.Entry<Integer, PendingStream> entry = pendingStreams.pollFirstEntry();
+            PendingStream pendingStream = entry.getValue();
+            pendingStream.sendFrames();
+            largestCreatedStreamId = pendingStream.streamId;
+        }
+    }
+
+    private void cancelPendingStreams() {
+        Exception e = new Exception("Connection closed.");
+        while (!pendingStreams.isEmpty()) {
+            PendingStream stream = pendingStreams.pollFirstEntry().getValue();
+            stream.close(e);
+        }
+    }
+
+    private void cancelGoAwayStreams(int lastStreamId, long errorCode, ByteBuf debugData) {
+        Iterator<PendingStream> iter = pendingStreams.values().iterator();
+        Exception e = new GoAwayException(lastStreamId, errorCode, debugData);
+        while (iter.hasNext()) {
+            PendingStream stream = iter.next();
+            if (stream.streamId > lastStreamId) {
+                iter.remove();
+                stream.close(e);
+            }
+        }
+    }
+
+    /**
+     * Determines whether or not we're allowed to create a new stream right now.
+     */
+    private boolean canCreateStream() {
+        Http2Connection.Endpoint<?> local = connection().local();
+        return (receivedSettings || local.numActiveStreams() < initialMaxConcurrentStreams)
```

Comment:
This actually changes the behavior, and I don't think in a good way. If the server does not specify MAX_CONCURRENT_STREAMS in the settings frame, then using localSettings makes the initial guess permanent. What the receivedSettings does is make the guess temporary until we get a settings from the server, at which point we know what the actual limit is (which may be infinite, because the setting was unspecified).